content/js1/sprints/1/prep/index.md

```diff
@@ -8,4 +8,443 @@ backlog= 'Module-JS1'
+
+{{</note>}}
+
+Once the computer evaluates these expressions the execution of the program is complete. But, we’re left with a problem. With the REPL, the user inputs an expression, the computer evaluates it and then prints the result. It then loops back and prompts the user for another input. But now the computer will go and execute each line sequentially until completion.
+
+So this new problem can be expressed as a question:
+
+> ❓ Problem
+>
+> "How can we check the values of variables in our program when it is being executed?"
+
+### 🖨️ Printing to the terminal
+
+In order to check the values when our program runs, we can make use of a function called `console.log`.  
+`console.log` will allow us to log expressions to the console while our program is executing. This is very useful for complex programs when we need to check the value of expressions in our program.
+
+Let's see how to use `console.log`
+
+In a file called `example.js`, we can write name of the function `console.log`.
+
+```js title="example.js"
+console.log;
+```
+
+However, if we run this file with Node, we won't be able to see anything in the terminal.
+As with `Math.round` we need to use the syntax for calling a function.
+So we can add parentheses after the function name to indicate we're calling this function:
+
+```js title="example.js"
+console.log("hello there!");
+```
+
+We should see the string "hello there!" logged out in the terminal.
+
+### Saving function output
+
+We can use variables to store the output from a function.
+
+Let’s consider the following file, `arithmetic.js`;
+
+```js title="arithmetic.js"
+const result = Math.round(10.3);
+```
+
+When this program is executed, it creates a variable called result and assigns to it **the output of the function**, in this case the rounded number.
+
+So `result` will store the value `10`.
+
+### 🔭 Logging vs function output
+
+Some functions output values that we can use in our program.
+
+`Math.round` takes an input, does a calculation and then outputs values that we can use when our program is running.
+
+However, some functions don't produce useful outputs in our running program; however they can still cause **effects**. Consider the following line:
+
+```js
+const result = console.log("hello world");
+```
+
+{{<note title="Exercise" type="exercise">}}
+
+Try predicting what `result` will evaluate to when the code above runs. Try executing this line in the Node REPL to observe what happens.
+
+{{</note>}}
+
+When this program runs, the variable `result` will evaluate to `undefined`.
+
+{{<note title="Note" type="note">}}
+
+`undefined` is a data type in JavaScript often used to represent the absence of a specific value
+
+{{</note>}}
+
+This is potentially confusing as `console.log` is a function with a set of instructions. `console.log` _does_ have an effect: it logs values to the console. However, `console.log` doesn't produce an output that we can use _inside_ the rest of our running program
```

Comment:
I would include a screenshot here of evaluating these two lines:
```console
> const result = console.log("hello world");
hello world
undefined
> result
undefined
> console.log(result)
undefined
undefined
```

and highlighting that the first and last evaluates result in _two_ lines of different colours, the side-effect print and the value evaluated to.

I'd maybe also go back to where we were introducing the REPL and include a similar screenshot just showing:

```console
> 5 + 10
15
```

and highlighting that `15` is the value evaluated by the expression, and maybe even reprise it here, so we have a consistent frame of reference to show the "there's an extra console.log output line" point.

I find that our trainees frequently get very confused between "what the REPL says a value evaluated to" and "what was console.log'd" because the difference is super subtle.

Reply:
My explanation was getting really long here towards the end here: I'm going to move this idea into a challenge and question in the coursework.